### PR TITLE
MAINT: Simplify ufunc pickling

### PR DESCRIPTION
**Pull request by seberg**
This also allows at least in principle numba dynamically
generated ufuncs to be pickled (with some hacking), see:

https://github.com/dask/distributed/issues/3450

If the name of the ufunc is set to a qualname, using this method,
pickle should be able to unpickle the ufunc correctly.
We may want to allow setting the module and qualname explicitly
on the ufunc object to remove the need for the custom pickler
completely.

---

We do have tests for rationals (which are much like scipy's ufuncs), and an old "pickle" (manually created string) which can still load with this.  I am not 100% sure on it, but to be honest, it seems cleaner to me to rely on pickle calling `whichmodule` rather than calling it ourselves, at least unless we want to actually store `__module__` and `__qualname__` on the ufunc itself, which I think may be a good idea.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
For reference, with this PR and https://github.com/numba/numba/pull/6234 I'm able to use numba's ufuncs on Dask Arrays as long as the ufunc used like

```python
>>> import mymodule
>>> mymodule.ufunc(dask_array).compute()
```

If you use `from mymodule import ufunc` dask still fails to deserialize it, but this is better than nothing.

File: numpy/core/__init__.py [hunk 1]
```diff
@@ -126,14 +125,17 @@ def _ufunc_reconstruct(module, name):
...
 def _ufunc_reduce(func):
-    from pickle import whichmodule
-    name = func.__name__
-    return _ufunc_reconstruct, (whichmodule(func, name), name)
+    # Report the `__name__`. pickle will try to find the module. Note that
+    # pickle supports for this `__name__` to be a `__qualname__`. It may
+    # make sense to add a `__qualname__` to ufuncs, to allow this more
+    # explicitly (Numba has ufuncs as attributes).
+    # See also: https://github.com/dask/distributed/issues/3450
+    return func.__name__
```
Comment (eric-wieser) [inline numpy/core/__init__.py:133]: Is this behavior of returning strings documented anywhere? I don't see it in https://docs.python.org/3/library/pickle.html.

**Comment by eric-wieser** [comment]
Is there a test that exercises this? A similar example failed for me:
```python
In [42]: class Wat:
    ...:     pass

In [44]: import copyreg

In [48]: def _wat_pickle(w):
    ...:     return "numpy.add"
    ...:

In [50]: copyreg.pickle(Wat, _wat_pickle)

In [51]: pickletools.dis(pickle.dumps(Wat()))
---------------------------------------------------------------------------
PicklingError                             Traceback (most recent call last)
<ipython-input-51-6471a402e037> in <module>
----> 1 pickletools.dis(pickle.dumps(Wat()))

PicklingError: Can't pickle <__main__.Wat object at 0x000002083E6719A0>: attribute lookup numpy.add on __main__ failed

```
Is something clever going on here where `pickle` tries the same module as the type, which is why it works for builtin ufuncs?

File: numpy/core/__init__.py (hunk 1 above)
Comment (seberg) [reply numpy/core/__init__.py:133]: Its a bit confusingly documented in [`copyreg`](https://docs.python.org/3/library/copyreg.html#copyreg.pickle)

There is also a note on the [`__reduce__` doctsring](https://docs.python.org/3/library/pickle.html#object.__reduce__) (second paragraph)

I would be happy to just implement `__reduce__` as well though, I guess I felt this is a start on that route.

**Comment by seberg** [comment]
You must not return the module, pickle does some magic to guess it (basically what we had, but I see no need for us to vendor it).

@eric-wieser yes there are tests, basic one here: https://github.com/numpy/numpy/blob/101192cbb5eedd8da97237a73ef2754fb9c65ddf/numpy/core/tests/test_ufunc.py#L182

Now, I had checked that there was a test for a "scipy" style ufunc. Not sure it actually is necessary (there is no real difference to ours), but it appears that the test might have been lost in history and is not present in master :/.

**Comment by seberg** [comment]
@eric-wieser I am a bit confused now why that fails with the error message as it does. What I tried was this (on this branch):
```
In [1]: import pickle                                                                                                   

In [2]: from scipy import special                                                                                       

In [3]: pickle.dumps(special.erf)                                                                                       
Out[3]: b'\x80\x04\x95!\x00\x00\x00\x00\x00\x00\x00\x8c\x15scipy.special._ufuncs\x94\x8c\x03erf\x94\x93\x94.
```
and it discovers the module correctly the same way as we previously did. Except that if the ufunc name was `something.attribute` it would actually add a `getattr()` call into the pickle.

**Review by eric-wieser — APPROVED** [review]
Looks fine, thanks for the explanations. I'm a bit puzzled now how to choose between `__reduce__` and `copyreg.pickle`, but the choice you make here seems consistent with many builtins.

**Comment by mattip** [comment]
> it appears that the test might have been lost in history

Can we revive that test or otherwise have a test that fails before, passes after this change?

File: numpy/core/tests/test_ufunc.py [hunk 2]
```diff
@@ -195,6 +199,15 @@ def test_pickle_withstring(self):
...
         assert_(pickle.loads(astring) is np.cos)
 
+    def test_pickle_name_is_qualname(self):
+        # This tests that a simplification of our ufunc pickle code will
+        # lead to allowing qualnames as names.  Future ufuncs should
+        # possible add a specific qualname, or a hook into pickling instead
+        # (dask+numba may benefit).
+        _pickleable_module_global.ufunc = umt._pickleable_module_global_ufunc
+        obj = pickle.loads(pickle.dumps(_pickleable_module_global.ufunc))
+        assert(obj is umt._pickleable_module_global_ufunc)
```
Comment (eric-wieser) [inline numpy/core/tests/test_ufunc.py:209]:
```suggestion
        assert obj is umt._pickleable_module_global_ufunc
```

**Review by eric-wieser — APPROVED** [review]

**Comment by mattip** [comment]
Thanks @seberg

**Pull request merged**